### PR TITLE
fix(types): add modifiersData types

### PR DESCRIPTION
**Pull request by atomiks**
- Can the`offset?:` type be DRY'd in Flow?
- Is there a non-null assertion in Flow? In TS: `state.modifiersData.popperOffsets!` asserts it exists (e.g. since we have `requires: ["popperOffsets"]`, we guarantee its availability)

### CODE REVIEW

**Comment by FezVrasta** [comment]
In Flow types are non-nullable and not optional by default. You need to use the `prop?: type` syntax to mark them as optional

**Comment by atomiks** [comment]
I'm referring to where I use it in our code. Previously we had "popperOffsets is always available" without a check, but now we need to check for its existence even though we guarantee it with the `requires` array.

**Comment by FezVrasta** [comment]
If the types make it optional, there’s no way to tell Flow to ignore the type. TS allows it because TS doesn’t give a sh*t about type soundness.

You would have to cast the type to `any` but I’d rather fix the code to handle the null case

File: src/types.js [hunk 1]
```diff
@@ -80,7 +80,39 @@ export type State = {|
...
+      centerOffset: number,
+    },
+    hide?: {
+      isReferenceHidden: boolean,
+      hasPopperEscaped: boolean,
+      referenceClippingOffsets: SideObject,
+      popperEscapeOffsets: SideObject,
+    },
+    offset?: {
+      auto: Offsets,
```
[the remark this answers is not in the document]
Comment (atomiks) [reply src/types.js:96]: Thank you! That's what I was asking in the first point, I knew there would be a simpler way to use the `Placements` type

**Comment by atomiks** [comment]
I'm getting a problem with `nanoid` dep after upgrading `flow-to-ts: ^0.1.6` so I think that's why the TS is failing as it's not building it properly. It should work otherwise.

**Comment by atomiks** [comment]
Nevermind, it seems like `flow-to-ts` can't handle the `[Placement]: Offsets` type. I've reverted to the literal types. Maybe open an issue on their repo later?

**Comment by FezVrasta** [comment]
There's already an open issue actually.

**Comment by FezVrasta** [comment]
Probably it would make sense to update the `Placement` type to use string literals, so that we can safely use it and have TS work too?

```js
type Placement = 
  | typeof top
  | 'top-start'
  | 'top-end'
  | typeof bottom
  | 'bottom-start'
  | 'bottom-end'
  | typeof right
  | 'right-start'
  | 'right-end'
  | typeof left
  | 'left-start'
  | 'left-end'
  | 'auto'
  | 'auto-start'
  | 'auto-end';
```

It's not very DRY, but better than having to inline the type.

**Comment by FezVrasta** [comment]
My suggestion doesn't seem to work 🙄

**Comment by FezVrasta** [comment]
I was thinking we could add some special comment syntax to tell flow-to-ts to use what's inside the comment.

So we could do:

```
/*:: type OffsetData = { [Placement]: Offsets } */ // this is just for Flow
/*;; type OffsetData = { [key in Placement]?: Offsets } */ // this would be just for TS
```

But I'll do that in a dedicated PR.

**Comment by FezVrasta** [comment]
Closing this in favor of https://github.com/popperjs/popper-core/pull/1079

**Pull request closed without merging**